### ISSUE TITLE
Question: Most Efficient and recommended Host Object Interaction Interface?

### ISSUE THREAD

**Issue by RickStrahl** [MicrosoftEdge/WebView2Feedback#823]
It looks like there are multiple ways to have JavaScript calls callback into .NET:

* Using the `postMessage()` API
* Using a registered Host object

I've been using Host Objects but the more I use it the more I feel like this is a very fickle interface that is bound to cause issues down the line. It works but inconsistencies in behavior, plus the current weird caching issues make me wonder how stable this will be in the long term.

By comparison the `postMessage` message interface is entirely string based and most likely will require serialization of everything plus make it harder to match existing interfaces that use the Web browser control. But since it's string based there are likely to be much fewer moving parts plus it looks like it doesn't require any COM under the cover, which likely also has quite a bit of overhead

**Question**

What do the designers recommend as the more efficient interface or the overall recommended interface?

Not knowing anything about how host object interop works for all I know it may be using the same message API as I doubt that the COM marshalling into JavaScript is native - there's likely some sort of message wrapping happening with dynamic COM proxies generated on the fly and I wonder if that is actually quite slow, compared even to passing messages via a string and serializing/deserializing on both ends.

So considerations are:

* Performance
* Stability

Any thoughts on this would be much appreciated.

**Comment by david-risney** [comment]
Great question!

## TLDR
AddHostObjectToScript vs PostWebMessage the performance should be similar, and for stability PostWebMessage is much simpler with fewer moving parts and so perhaps better in that regard if you only care about sending raw messages.

## Details
I should also point out that this is all regarding our *current* implementation. I'm not guaranteeing we won't change how this is implemented in the future. Additionally, the discussion of relative performance or stability is based on what I would expect given the design of the features and not based on any actual statistics.

AddHostObjectToScript has a native side in the host app and a JavaScript side in the renderer process that communicate via mojo (chromium's cross platform IPC mechanism). JavaScript in the renderer process implements proxy objects that translate method calls into mojo messages and translate mojo message responses into method call results. Similarly on the native side the WebView2 code interprets the mojo messages and makes calls on the IDispatch objects provided by the end developer or generated as a result of a previous JavaScript call. This design should let us replace IDispatch with other sorts of languages or runtimes even on other platforms if need be in the future.

PostWebMessage\* uses mojo to communicate back and forth between the native WebView2 side and the JavaScript in the renderer.

Both mechanisms will have to do some amount of translation between JSON strings and JavaScript objects and should have similar perf for communicating messages back and forth. AddHostObjectToScript has much more code for dealing with IDispatch and creating JavaScript proxies. I wouldn't expect the perf impact of that overhead to be large but it is definitely the more complicated feature. Using PostWebMessage pushes handling those kind of complications to the end developer. If your scenario only involves sending simple messages then PostWebMessage is the easier and better option.
